{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome Detective output conversion to CAMI profiling\n",
    "\n",
    "Author: Sam Nooij  \n",
    "Date: 2018-05-24\n",
    "\n",
    "\n",
    "Input: ## what exactly do I need...?\n",
    "- sample ID\n",
    "- taxon\n",
    "- number of reads\n",
    "- total reads\n",
    "\n",
    "    - These data (for both the assignments and the discoveries) can be derived from: the \"bokeh input\", output by `GenomeDetective_heatmaps.py`\n",
    "\n",
    "- NCBI taxonomy DB (through ETE toolkit, see [this tutorial](http://etetoolkit.org/docs/2.3/tutorial/tutorial_ncbitaxonomy.html))\n",
    "\n",
    "Output: table in [CAMI profiling format](https://github.com/bioboxes/rfc/blob/60263f34c57bc4137deeceec4c68a7f9f810f6a5/data-format/profiling.mkd)\n",
    "\n",
    "Required python packages:\n",
    " - pandas\n",
    " - ete3\n",
    " \n",
    "For automatic use in snakemake. The corresponding snakemake rule should provide the input:\n",
    " - the parsed report file (\"results/GenomeDetective_results.csv\")\n",
    " - a name for the output (e.g. \"results/GenomeDetective_CAMI-profiling.tsv\")\n",
    " \n",
    "  ** Remember that an output has to be generated for each sample, separately! **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import required python libraries\n",
    "import pandas as pd         #dataframe and csv export\n",
    "from ete3 import NCBITaxa   #work with NCBI taxonomy\n",
    "\n",
    "#DATA_TABLE = snakemake.input[0]\n",
    "#PROFILES = snakemake.output[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# debugging cell:\n",
    "DATA_TABLE = \"../tmp/bokeh_input.csv\"\n",
    "PROFILE = \"../results/GenomeDetective_CAMI-profiling.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      TAXID     RANK                                        TAXPATH  \\\n",
      "0    142786    genus                10239|439488|35278|11974|142786   \n",
      "1   1211417  species                     10239|12333|156614|1211417   \n",
      "2     36343  species           10239|35237|28883|10699|186532|36343   \n",
      "3   1623305    genus                10239|35237|28883|10699|1623305   \n",
      "4   1105171  species         10239|35237|28883|10699|196894|1105171   \n",
      "5     31537  species           10239|35237|28883|10699|186532|31537   \n",
      "6   1327037  species         10239|35237|28883|10699|196894|1327037   \n",
      "7   1655644  species               10239|29258|10841|117574|1655644   \n",
      "8   1176422  species         10239|35237|28883|10699|196894|1176422   \n",
      "9   1921554  species        10239|35237|28883|10699|1921553|1921554   \n",
      "10  1921555  species        10239|35237|28883|10699|1921553|1921555   \n",
      "11  1734382  species         10239|35237|28883|10744|196895|1734382   \n",
      "12  1676182  species               10239|29258|10841|117574|1676182   \n",
      "13   198538  species  10239|35237|28883|10744|196895|1758253|198538   \n",
      "14  1623305    genus                10239|35237|28883|10699|1623305   \n",
      "15  1720493  species         10239|35237|28883|10699|196894|1720493   \n",
      "16    10756  species    10239|35237|28883|10744|542836|186846|10756   \n",
      "17  1305708  species         10239|35237|28883|10699|196894|1305708   \n",
      "18  1981026  species        10239|35237|28883|10699|1623289|1981026   \n",
      "\n",
      "                                            TAXPATHSN  PERCENTAGE  \n",
      "0   Viruses|ssRNA viruses|ssRNA positive-strand vi...   63.887417  \n",
      "1   Viruses|unclassified bacterial viruses|environ...    9.295839  \n",
      "2   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.108443  \n",
      "3   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.079242  \n",
      "4   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.035912  \n",
      "5   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.032027  \n",
      "6   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.012716  \n",
      "7   Viruses|ssDNA viruses|Microviridae|unclassifie...    0.009773  \n",
      "8   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.007300  \n",
      "9   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.005652  \n",
      "10  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.005652  \n",
      "11  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.004239  \n",
      "12  Viruses|ssDNA viruses|Microviridae|unclassifie...    0.004239  \n",
      "13  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.002119  \n",
      "14  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.079242  \n",
      "15  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.001295  \n",
      "16  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.001060  \n",
      "17  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000589  \n",
      "18  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000235  \n",
      "      TAXID     RANK                                       TAXPATH  \\\n",
      "0    142786    genus               10239|439488|35278|11974|142786   \n",
      "1     12239  species         10239|439488|35278|675071|12234|12239   \n",
      "2   1187973  species      10239|35325|585893|104394|573164|1187973   \n",
      "3   1623305    genus               10239|35237|28883|10699|1623305   \n",
      "4     10778  species   10239|35237|28883|10744|542836|186846|10778   \n",
      "5   1309583  species        10239|35237|28883|10744|196895|1309583   \n",
      "6    338473  species  10239|35237|28883|10744|542836|545935|338473   \n",
      "7     35241  species          10239|35237|28883|10699|196894|35241   \n",
      "8    151536  species         10239|35237|28883|10699|196894|151536   \n",
      "9   1161930  species        10239|35237|28883|10699|196894|1161930   \n",
      "10    37105  species          10239|35237|28883|10699|196894|37105   \n",
      "\n",
      "                                            TAXPATHSN  PERCENTAGE  \n",
      "0   Viruses|ssRNA viruses|ssRNA positive-strand vi...    0.635983  \n",
      "1   Viruses|ssRNA viruses|ssRNA positive-strand vi...    0.018020  \n",
      "2   Viruses|dsRNA viruses|Picobirnaviridae|Picobir...    0.324017  \n",
      "3   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.510875  \n",
      "4   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.172855  \n",
      "5   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.105595  \n",
      "6   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.017217  \n",
      "7   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.015495  \n",
      "8   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.005739  \n",
      "9   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000689  \n",
      "10  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000230  \n",
      "     TAXID     RANK                                         TAXPATH  \\\n",
      "0   142786    genus                 10239|439488|35278|11974|142786   \n",
      "1  1803956  species  10239|439488|35278|464095|12058|138954|1803956   \n",
      "2  2003400    genus          10239|439488|35278|76804|76803|2003400   \n",
      "3   138949  species    10239|439488|35278|464095|12058|12059|138949   \n",
      "4   687331    genus                       10239|29258|687329|687331   \n",
      "5   186789    genus                  10239|35237|28883|10662|186789   \n",
      "6    10742  species           10239|35237|28883|10699|2169654|10742   \n",
      "7  1815968  species   10239|35237|28883|10662|186789|329160|1815968   \n",
      "\n",
      "                                           TAXPATHSN  PERCENTAGE  \n",
      "0  Viruses|ssRNA viruses|ssRNA positive-strand vi...   28.238274  \n",
      "1  Viruses|ssRNA viruses|ssRNA positive-strand vi...   10.480651  \n",
      "2  Viruses|ssRNA viruses|ssRNA positive-strand vi...    0.513393  \n",
      "3  Viruses|ssRNA viruses|ssRNA positive-strand vi...    0.014144  \n",
      "4  Viruses|ssDNA viruses|Anelloviridae|Alphatorqu...    0.000718  \n",
      "5  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.011885  \n",
      "6  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.006912  \n",
      "7  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000546  \n",
      "      TAXID     RANK                                            TAXPATH  \\\n",
      "0    138948  species       10239|439488|35278|464095|12058|12059|138948   \n",
      "1    142786    genus                    10239|439488|35278|11974|142786   \n",
      "2   1803956  species     10239|439488|35278|464095|12058|138954|1803956   \n",
      "3    138949  species       10239|439488|35278|464095|12058|12059|138949   \n",
      "4   1982579  species             10239|35237|28883|10744|542960|1982579   \n",
      "5     12345  species        10239|35237|28883|10744|542836|186846|12345   \n",
      "6   1124654  species     10239|35237|28883|10744|542960|1231360|1124654   \n",
      "7   1204535  species             10239|35237|28883|10662|196896|1204535   \n",
      "8    338473  species       10239|35237|28883|10744|542836|545935|338473   \n",
      "9     10756  species        10239|35237|28883|10744|542836|186846|10756   \n",
      "10  1981167  species    10239|35237|28883|10744|1981156|1981159|1981167   \n",
      "11  1150869  species  10239|35237|28883|10744|1981156|1981159|210846...   \n",
      "\n",
      "                                            TAXPATHSN  PERCENTAGE  \n",
      "0   Viruses|ssRNA viruses|ssRNA positive-strand vi...   60.284583  \n",
      "1   Viruses|ssRNA viruses|ssRNA positive-strand vi...   12.803462  \n",
      "2   Viruses|ssRNA viruses|ssRNA positive-strand vi...    4.674955  \n",
      "3   Viruses|ssRNA viruses|ssRNA positive-strand vi...    0.012231  \n",
      "4   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.055884  \n",
      "5   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.024014  \n",
      "6   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.021230  \n",
      "7   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.015662  \n",
      "8   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.009795  \n",
      "9   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.004972  \n",
      "10  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000994  \n",
      "11  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000398  \n"
     ]
    }
   ],
   "source": [
    "results_df = pd.read_csv(DATA_TABLE)\n",
    "samples = sorted(set(results_df[\"sample\"]))\n",
    "\n",
    "set_taxonomy = \"superkingdom|Baltimore|phylum|class|order|family|genus|species\"\n",
    "taxonomy_order = set_taxonomy.split('|')\n",
    "baltimore_taxa = {35237 : \"Baltimore\", #dsDNA, no RNA stage (class I)\n",
    "                  2080735 : \"Baltimore\", #DNA viruses (separate group made for an Archaeal virus, should be class I)\n",
    "                  29258 : \"Baltimore\", #ssDNA viruses (class II)\n",
    "                  35325 : \"Baltimore\", #dsRNA viruses (class III)\n",
    "                  35278 : \"Baltimore\", #ssRNA positive-strand viruses (class IV)\n",
    "                  35301 : \"Baltimore\", #ssRNA negative-strand viruses (class V)\n",
    "                  35268 : \"Baltimore\", #Retro-transcribing viruses, dsDNA-RT viruses (class VII)                             \n",
    "                  }\n",
    "\n",
    "#Retroviruses have no annotated Baltimore class in NCBI Taxonomy\n",
    "#if order == 2169561 or order == \"Ortervirales\":\n",
    "#    baltimore = \"ssRNA-RT viruses\" #(class VI)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "    I: dsDNA viruses (e.g. Adenoviruses, Herpesviruses, Poxviruses)\n",
    "    II: ssDNA viruses (+ strand or \"sense\") DNA (e.g. Parvoviruses)\n",
    "    III: dsRNA viruses (e.g. Reoviruses)\n",
    "    IV: (+)ssRNA viruses (+ strand or sense) RNA (e.g. Picornaviruses, Togaviruses)\n",
    "    V: (−)ssRNA viruses (− strand or antisense) RNA (e.g. Orthomyxoviruses, Rhabdoviruses)\n",
    "    VI: ssRNA-RT viruses (+ strand or sense) RNA with DNA intermediate in life-cycle (e.g. Retroviruses)\n",
    "    VII: dsDNA-RT viruses DNA with RNA intermediate in life-cycle (e.g. Hepadnaviruses)\n",
    "\n",
    "source: https://en.wikipedia.org/wiki/Baltimore_classification\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "PROFILES = [ \"%s/%s_%s\" % ('/'.join(PROFILE.split('/')[:-1]), sample, PROFILE.split('/')[-1]) for sample in samples ]\n",
    "\n",
    "ncbi = NCBITaxa()\n",
    "#In case you need to update, uncomment the next line:\n",
    "#ncbi.update_taxonomy_database()\n",
    "\n",
    "#Create separate files for each sample\n",
    "for sample in samples:\n",
    "    #Pick the right (pro)file name, that matches the sample\n",
    "    profile = [ p for p in PROFILES if sample in p ][0]\n",
    "    \n",
    "    subset = results_df[results_df[\"sample\"] == sample]\n",
    "    taxa = subset[\"Assignment\"]\n",
    "    total_percentages = subset[\"percentage_of_total_reads\"]\n",
    "    \n",
    "    rank_list_list = []\n",
    "    output_list = []\n",
    "    taxid_list = []\n",
    "    ranks = []\n",
    "    taxpath_list = []\n",
    "    taxpathsn_list = []\n",
    "    percentage_list = []\n",
    "    \n",
    "    for name in taxa:\n",
    "        #remove names that have some addition in brackets,\n",
    "        # like \" (segment 1)\"\n",
    "        if ' (' in name:\n",
    "            original_name = name\n",
    "            name = name[:name.index(' (')]\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        taxon_and_id = ncbi.get_name_translator([name])\n",
    "        #ncbi.get_name_translator() returns a dictionary { 'taxon' : [id]}\n",
    "        taxid = taxon_and_id[name]\n",
    "        #taxid is a list with one number\n",
    "        taxid_nr = taxid[0]\n",
    "        taxid_list.append(taxid_nr)\n",
    "\n",
    "        rank_dict = ncbi.get_rank(taxid)\n",
    "        #ncbi.get_rank() requires a list of IDs, and returns a dictionary:\n",
    "        # {id: 'rank'}\n",
    "        rank = rank_dict[taxid_nr]\n",
    "        ranks.append(rank)\n",
    "\n",
    "        tax_path_dict = ncbi.get_lineage_translator(taxid)#[taxid_nr]\n",
    "        #ncbi.get_lineage_translator() requires a list of IDs, and returns\n",
    "        # a dictionary {leaf_id: [root_id, node_id, leaf_id]}\n",
    "        tax_path = tax_path_dict[taxid_nr][1:]\n",
    "\n",
    "        tax_path_sn = []\n",
    "        #with a for-loop you can translate the taxids in the list\n",
    "        # 'tax_path' to their corresponding scientific names (sn)\n",
    "        for t in tax_path:\n",
    "            tax_path_sn.append(ncbi.get_taxid_translator([t])[t])\n",
    "\n",
    "        rank_list = []\n",
    "        #Making this list requires using a for-loop;\n",
    "        # using the function on a list makes an UNORDERED dictionary\n",
    "        #Also, since the path differs between branches, I will look\n",
    "        # for the longest using a list of lists\n",
    "        for taxid in tax_path:\n",
    "            rank_dict = ncbi.get_rank([taxid])\n",
    "            rank = rank_dict[taxid]\n",
    "            rank_list.append(rank)\n",
    "\n",
    "        rank_list_list.append(rank_list)    \n",
    "\n",
    "        tax_path_string = '|'.join(map(str, tax_path))\n",
    "        tax_path_sn_string = '|'.join(tax_path_sn)\n",
    "        \n",
    "        taxpath_list.append(tax_path_string)\n",
    "        taxpathsn_list.append(tax_path_sn_string)\n",
    "        \n",
    "        try:\n",
    "            percentage = subset.loc[subset[\"Assignment\"] == name][\"percentage_of_total_reads\"].values[0]\n",
    "        except:\n",
    "            percentage = subset.loc[subset[\"Assignment\"] == original_name][\"percentage_of_total_reads\"].values[0]\n",
    "\n",
    "        percentage_list.append(percentage)\n",
    "            \n",
    "        output_line = \"%s\\t%s\\t%s\\t%s\\t%s\" % (taxid_nr, rank, tax_path_string, tax_path_sn_string, percentage)\n",
    "        \n",
    "        output_list.append(output_line)\n",
    "        output_dict = { \"TAXID\" : taxid_list, \"RANK\" : ranks, \"TAXPATH\" : taxpath_list,\n",
    "                      \"TAXPATHSN\" : taxpathsn_list, \"PERCENTAGE\" : percentage_list}\n",
    "        \n",
    "    longest_taxonomy = '|'.join(max(rank_list_list, key = len))\n",
    "    \n",
    "    #Read the specification for details about this header:\n",
    "    #https://github.com/bioboxes/rfc/blob/60263f34c57bc4137deeceec4c68a7f9f810f6a5/data-format/profiling.mkd\n",
    "    header = \"\"\"# Taxonomic Profiling Output\n",
    "@SampleID:%s\n",
    "@Version:0.9.3\n",
    "@Ranks:%s\\t#the longest path in this sample: virus taxonomy is messy\n",
    "@TaxonomyID:ncbi-taxonomy_2018-05-25\n",
    "@@TAXID\\tRANK\\tTAXPATH\\tTAXPATHSN\\tPERCENTAGE\n",
    "\"\"\" % (sample, longest_taxonomy)\n",
    "    \n",
    "    #with open(profile, 'w') as output_table:\n",
    "    #    output_table.write(header)\n",
    "    #    output_table.write('\\n'.join(output_list))\n",
    "    output_df = pd.DataFrame(output_dict)\n",
    "    print(output_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'superkingdom': 0, 'Baltimore': 1, 'phylum': 2, 'class': 3, 'order': 4, 'family': 5, 'genus': 6, 'species': 7}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/software/python/anaconda2/envs/genomedetective/lib/python3.6/site-packages/ipykernel_launcher.py:9: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TAXID</th>\n",
       "      <th>RANK</th>\n",
       "      <th>TAXPATH</th>\n",
       "      <th>TAXPATHSN</th>\n",
       "      <th>PERCENTAGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>10239</td>\n",
       "      <td>superkingdom</td>\n",
       "      <td>10239</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>77.908180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>35278</td>\n",
       "      <td>Baltimore</td>\n",
       "      <td>10239|439488|35278</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>77.775231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35237</td>\n",
       "      <td>Baltimore</td>\n",
       "      <td>10239|35237</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage</td>\n",
       "      <td>0.132949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>464095</td>\n",
       "      <td>order</td>\n",
       "      <td>10239|439488|35278|464095</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>64.971769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>28883</td>\n",
       "      <td>order</td>\n",
       "      <td>10239|35237|28883</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudovirales</td>\n",
       "      <td>0.132949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>12058</td>\n",
       "      <td>family</td>\n",
       "      <td>10239|439488|35278|464095|12058</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>64.971769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>11974</td>\n",
       "      <td>family</td>\n",
       "      <td>10239|439488|35278|11974</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>12.803462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>10744</td>\n",
       "      <td>family</td>\n",
       "      <td>10239|35237|28883|10744</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.117288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>10662</td>\n",
       "      <td>family</td>\n",
       "      <td>10239|35237|28883|10662</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.015662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>12059</td>\n",
       "      <td>genus</td>\n",
       "      <td>10239|439488|35278|464095|12058|12059</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>60.296814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>142786</td>\n",
       "      <td>genus</td>\n",
       "      <td>10239|439488|35278|11974|142786</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>12.803462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>138954</td>\n",
       "      <td>genus</td>\n",
       "      <td>10239|439488|35278|464095|12058|138954</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>4.674955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>542960</td>\n",
       "      <td>genus</td>\n",
       "      <td>10239|35237|28883|10744|542960</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.077115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>186846</td>\n",
       "      <td>genus</td>\n",
       "      <td>10239|35237|28883|10744|542836|186846</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.028986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>1981159</td>\n",
       "      <td>genus</td>\n",
       "      <td>10239|35237|28883|10744|1981156|1981159</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.001392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>138948</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|439488|35278|464095|12058|12059|138948</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>60.284583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1803956</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|439488|35278|464095|12058|138954|1803956</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>4.674955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1982579</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|35237|28883|10744|542960|1982579</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.055884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>12345</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|35237|28883|10744|542836|186846|12345</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.024014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1124654</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|35237|28883|10744|542960|1231360|1124654</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.021230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1204535</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|35237|28883|10662|196896|1204535</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.015662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>138949</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|439488|35278|464095|12058|12059|138949</td>\n",
       "      <td>Viruses|ssRNA viruses|ssRNA positive-strand vi...</td>\n",
       "      <td>0.012231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>338473</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|35237|28883|10744|542836|545935|338473</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.009795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10756</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|35237|28883|10744|542836|186846|10756</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.004972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1981167</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|35237|28883|10744|1981156|1981159|1981167</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.000994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1150869</td>\n",
       "      <td>species</td>\n",
       "      <td>10239|35237|28883|10744|1981156|1981159|210846...</td>\n",
       "      <td>Viruses|dsDNA viruses, no RNA stage|Caudoviral...</td>\n",
       "      <td>0.000398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      TAXID          RANK                                            TAXPATH  \\\n",
       "14    10239  superkingdom                                              10239   \n",
       "24    35278     Baltimore                                 10239|439488|35278   \n",
       "35    35237     Baltimore                                        10239|35237   \n",
       "23   464095         order                          10239|439488|35278|464095   \n",
       "39    28883         order                                  10239|35237|28883   \n",
       "28    12058        family                    10239|439488|35278|464095|12058   \n",
       "22    11974        family                           10239|439488|35278|11974   \n",
       "25    10744        family                            10239|35237|28883|10744   \n",
       "27    10662        family                            10239|35237|28883|10662   \n",
       "15    12059         genus              10239|439488|35278|464095|12058|12059   \n",
       "1    142786         genus                    10239|439488|35278|11974|142786   \n",
       "31   138954         genus             10239|439488|35278|464095|12058|138954   \n",
       "41   542960         genus                     10239|35237|28883|10744|542960   \n",
       "44   186846         genus              10239|35237|28883|10744|542836|186846   \n",
       "43  1981159         genus            10239|35237|28883|10744|1981156|1981159   \n",
       "0    138948       species       10239|439488|35278|464095|12058|12059|138948   \n",
       "2   1803956       species     10239|439488|35278|464095|12058|138954|1803956   \n",
       "4   1982579       species             10239|35237|28883|10744|542960|1982579   \n",
       "5     12345       species        10239|35237|28883|10744|542836|186846|12345   \n",
       "6   1124654       species     10239|35237|28883|10744|542960|1231360|1124654   \n",
       "7   1204535       species             10239|35237|28883|10662|196896|1204535   \n",
       "3    138949       species       10239|439488|35278|464095|12058|12059|138949   \n",
       "8    338473       species       10239|35237|28883|10744|542836|545935|338473   \n",
       "9     10756       species        10239|35237|28883|10744|542836|186846|10756   \n",
       "10  1981167       species    10239|35237|28883|10744|1981156|1981159|1981167   \n",
       "11  1150869       species  10239|35237|28883|10744|1981156|1981159|210846...   \n",
       "\n",
       "                                            TAXPATHSN  PERCENTAGE  \n",
       "14                                            Viruses   77.908180  \n",
       "24  Viruses|ssRNA viruses|ssRNA positive-strand vi...   77.775231  \n",
       "35                Viruses|dsDNA viruses, no RNA stage    0.132949  \n",
       "23  Viruses|ssRNA viruses|ssRNA positive-strand vi...   64.971769  \n",
       "39   Viruses|dsDNA viruses, no RNA stage|Caudovirales    0.132949  \n",
       "28  Viruses|ssRNA viruses|ssRNA positive-strand vi...   64.971769  \n",
       "22  Viruses|ssRNA viruses|ssRNA positive-strand vi...   12.803462  \n",
       "25  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.117288  \n",
       "27  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.015662  \n",
       "15  Viruses|ssRNA viruses|ssRNA positive-strand vi...   60.296814  \n",
       "1   Viruses|ssRNA viruses|ssRNA positive-strand vi...   12.803462  \n",
       "31  Viruses|ssRNA viruses|ssRNA positive-strand vi...    4.674955  \n",
       "41  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.077115  \n",
       "44  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.028986  \n",
       "43  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.001392  \n",
       "0   Viruses|ssRNA viruses|ssRNA positive-strand vi...   60.284583  \n",
       "2   Viruses|ssRNA viruses|ssRNA positive-strand vi...    4.674955  \n",
       "4   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.055884  \n",
       "5   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.024014  \n",
       "6   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.021230  \n",
       "7   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.015662  \n",
       "3   Viruses|ssRNA viruses|ssRNA positive-strand vi...    0.012231  \n",
       "8   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.009795  \n",
       "9   Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.004972  \n",
       "10  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000994  \n",
       "11  Viruses|dsDNA viruses, no RNA stage|Caudoviral...    0.000398  "
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_taxa = set('|'.join(output_df[\"TAXPATH\"]).split('|'))\n",
    "#Take all rows in column \"TAXPATH\", join them together with '|',\n",
    "# split them into a list and take the uniques with 'set'\n",
    "\n",
    "rows = output_df.shape[0]\n",
    "new_row = rows\n",
    "\n",
    "for t in all_taxa:\n",
    "    percentages = output_df[0:rows][output_df[\"TAXPATH\"].str.contains(t)][\"PERCENTAGE\"]\n",
    "    perc_sum = sum(percentages)\n",
    "    rank = ncbi.get_rank([t])[int(t)]\n",
    "    tax_path = ncbi.get_lineage_translator([t])[int(t)][1:]\n",
    "    tax_path_sn = []\n",
    "    #with a for-loop you can translate the taxids in the list\n",
    "    # 'tax_path' to their corresponding scientific names (sn)\n",
    "    for tax in tax_path:\n",
    "        tax_path_sn.append(ncbi.get_taxid_translator([tax])[tax])\n",
    "        \n",
    "    if rank == \"no rank\":\n",
    "        if int(t) in baltimore_taxa:\n",
    "            rank = baltimore_taxa[int(t)]\n",
    "        \n",
    "    new_row += 1\n",
    "    if rank in taxonomy_order:\n",
    "        output_df.loc[new_row] = [int(t), rank, '|'.join(map(str, tax_path)), '|'.join(tax_path_sn), perc_sum]\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "### SORTER THANKS TO: https://stackoverflow.com/a/23483221\n",
    "taxonomy_sorter_index = dict(zip(taxonomy_order, range(len(taxonomy_order))))\n",
    "print(taxonomy_sorter_index)\n",
    "\n",
    "output_df['taxonomy_sorter'] = output_df[\"RANK\"].map(taxonomy_sorter_index)\n",
    "\n",
    "output_df.sort_values(by = [\"taxonomy_sorter\", \"PERCENTAGE\"], ascending = [True, False], inplace = True)\n",
    "\n",
    "output_df.drop('taxonomy_sorter', 1, inplace = True)\n",
    "###\n",
    "\n",
    "output_df.drop_duplicates(keep = \"first\", inplace = True)\n",
    "\n",
    "output_df\n",
    "\n",
    "#Test: find and summarise all Picornaviridae:\n",
    "#picorna_percentages = output_df[output_df[\"TAXPATH\"].str.contains('12058')][\"PERCENTAGE\"]\n",
    "#picorna_sum = sum(picorna_percentages)\n",
    "#picorna_rank = ncbi.get_rank([12058])[12058]\n",
    "\n",
    "#print(picorna_percentages, \"\\n\", round(picorna_sum, 3), picorna_rank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df.to_csv(\"../tmp/20180622_Test_new_profile.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
