{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome Detective heatmap creation script\n",
    "\n",
    "Author: Sam Nooij  \n",
    "Date: 2018-05-22\n",
    "\n",
    "Input:\n",
    "1. parsed XML file (e.g. \"tmp/GenomeDetective_results.csv\") as generated by bin/GenomeDetective_XML_parser.py\n",
    "2. result CSV files as generated by [GenomeDetective](http://www.genomedetective.com/app/typingtool/virus/)\n",
    "    a. Assignments\n",
    "    b. Discoveries\n",
    "\n",
    "Output: table in csv format, providing:\n",
    "- run_id\n",
    "- sample_id\n",
    "- taxon\n",
    "- total_reads\n",
    "- viral_reads \n",
    "- coverage%\n",
    "- contigs\n",
    "- number_of_reads\n",
    "- fraction_of_total_reads\n",
    "- percentage_of_total\n",
    "- fraction_of_viral_reads\n",
    "- percentage_of_viral\n",
    "\n",
    "Interactive (HTML) heatmaps of:\n",
    "- assigned taxa\n",
    "- discovered taxa\n",
    "- assigned and discovered taxa (together in the same map)\n",
    "\n",
    "Required python packages:\n",
    " - pandas\n",
    " - bokeh\n",
    " \n",
    "For automatic use in snakemake. The corresponding snakemake rule should provide the input:\n",
    " - the parsed XML file (\"tmp/GenomeDetective_results.csv\")\n",
    " - a list of CSV files (their names, as strings; e.g. [ \"1_a_results.csv\", \"1_b_results.csv\" ]; also for the discovery)\n",
    " - a name for the output files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import all required libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from bokeh.plotting import figure, show, output_file\n",
    "from bokeh.models import HoverTool, ColumnDataSource\n",
    "from bokeh.io import output_notebook\n",
    "from sys import argv\n",
    "\n",
    "#You may set the colour by passing an argument to this script\n",
    "if len(argv) > 1:\n",
    "    COLOUR = list(str(argv[1]))\n",
    "#If you don't, the default (brown) is used\n",
    "else:\n",
    "    COLOUR = [\"#6b2d18\"] #Selected from coffee beans: http://s.eatthis-cdn.com/media/images/ext/851818315/coffee-beans.jpg\n",
    "\n",
    "#ASSIGNMENTS = snakemake.input['assignments']\n",
    "#DISCOVERIES = snakemake.input['discoveries']\n",
    "#PARSED_XML = snakemake.input['parsed_xml']\n",
    "#MAP_A = snakemake.input['heatmap_a']\n",
    "#MAP_D = snakemake.input['heatmap_d']\n",
    "#MAP_AD = snakemake.input['heatmap_ad']\n",
    "#OUTPUT_FILE = snakemake.output[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### code block for developing/debugging only! ###\n",
    "import glob\n",
    "\n",
    "ASSIGNMENTS = glob.glob(\"../results/*_results.csv\")\n",
    "DISCOVERIES = glob.glob(\"../results/*_discovery.csv\")\n",
    "PARSED_XML = \"../tmp/GenomeDetective_results-xml.csv\"\n",
    "MAP_A = \"../tmp/Test_A_heatmap.html\"\n",
    "MAP_D = \"../tmp/Test_D_heatmap.html\"\n",
    "MAP_AD = \"../tmp/Test_AD_heatmap.html\"\n",
    "OUTPUT_FILE = \"../tmp/Bokeh_heatmap_data.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Functions for parsing, dataframe building and heatmap creation\n",
    "def pull_sample_name(filename):\n",
    "    \"\"\"\n",
    "    The sample and run IDs are in the filename, e.g.:\n",
    "    \"3_10_results.xml\"\n",
    "    where the 3 is the run ID, and the 10 is the sample ID\n",
    "    \"\"\"\n",
    "    error_msg = \"\"\"\n",
    "Expected underscores in the filename with the sample ID, e.g.\n",
    "3_1_results.xml\n",
    "Please provide sample names in this format.\n",
    "    \"\"\"\n",
    "    \n",
    "    assert filename.count('_') >= 2, \\\n",
    "        \"%s\" % error_msg\n",
    "        \n",
    "    run_id = filename.split('/')[-1].split('_')[0]\n",
    "    \n",
    "    if filename.count('_') > 2:\n",
    "        sample_id = '_'.join(filename.split('/')[-1].split('_')[1:-1])\n",
    "    else:\n",
    "        sample_id = filename.split('/')[-1].split('_')[1]\n",
    "    \n",
    "    return run_id, sample_id\n",
    "\n",
    "def create_concatenated_dataframe(csv_list):\n",
    "    \"\"\"\n",
    "    Input: a list of Genome Detective output CSV files,\n",
    "          e.g. [\"3_1_results.csv\", \"4_D_results.csv\"]\n",
    "    Output: One concatenated dataframe of all the input files\n",
    "    \"\"\"\n",
    "    csv_list = sorted(csv_list)\n",
    "    \n",
    "        \n",
    "    #Step 2: open the files as dataframe, remove \"Contigs\" column and add sample IDs\n",
    "    df_list = []\n",
    "    for results_file in csv_list:\n",
    "        results_df = pd.read_csv(results_file)\n",
    "        results_df = results_df.drop(\"Contigs\", axis = 1) #remove unnecessary (and long!) column\n",
    "        run_id, sample_id = pull_sample_name(results_file)\n",
    "        results_df[\"run_id\"] = run_id\n",
    "        results_df[\"sample_id\"] = sample_id\n",
    "        df_list.append(results_df)\n",
    "\n",
    "    #Step 3: concatenate the dataframes\n",
    "    super_df = pd.concat(df_list, ignore_index=True)\n",
    "    \n",
    "    return(super_df)\n",
    "\n",
    "def calculate_fractions(dataframe):\n",
    "    \"\"\"\n",
    "    Input: Dataframe with columns \"number_of_reads\",\n",
    "        \"total_reads\", and \"viral_reads\"\n",
    "    Output: Dataframe with fractiond and percentages\n",
    "    \"\"\"\n",
    "    dataframe = dataframe.assign(fraction_of_total_reads = lambda dataframe:\n",
    "                          dataframe[\"Mapped # Reads\"] / dataframe[\"total_reads\"])\n",
    "\n",
    "    dataframe = dataframe.assign(fraction_of_viral_reads = lambda dataframe:\n",
    "                          dataframe[\"Mapped # Reads\"] / dataframe[\"viral_reads\"])\n",
    "\n",
    "    dataframe = dataframe.assign(percentage_of_total_reads = lambda dataframe:\n",
    "                          dataframe[\"fraction_of_total_reads\"] * 100)\n",
    "\n",
    "    dataframe = dataframe.assign(percentage_of_viral_reads = lambda dataframe:\n",
    "                          dataframe[\"fraction_of_viral_reads\"] * 100)\n",
    "\n",
    "    return(dataframe)\n",
    "\n",
    "def create_heatmaps(dataframe):\n",
    "    \"\"\"\n",
    "    Input: Dataframe with all required data:\n",
    "        Taxon, number of reads, assigned/discovered\n",
    "    Output: 3 heatmaps (assigned/discovered/both), as html files\n",
    "            csv file with the data\n",
    "    \"\"\"\n",
    "    \n",
    "    def create_heatmap(subset_df, title, filename):\n",
    "        samples = subset_df[\"sample\"]\n",
    "        assigned = subset_df[\"Assignment\"]\n",
    "        reads = subset_df[\"Mapped # Reads\"]\n",
    "        total_reads = subset_df[\"total_reads\"]\n",
    "        viral_reads = subset_df[\"viral_reads\"]\n",
    "        percent_of_total = subset_df[\"percentage_of_total_reads\"]\n",
    "        percent_of_viral = subset_df[\"percentage_of_viral_reads\"]\n",
    "        contigs = subset_df[\"# Contigs\"]\n",
    "        coverage = subset_df[\"Coverage (%)\"]\n",
    "\n",
    "        colors = len(samples) * COLOUR #multiply to make an equally long list\n",
    "        colors = colors[:len(samples)]\n",
    "        \n",
    "        max_load = max(percent_of_total)\n",
    "        alphas = [ min( x / float(max_load), 0.9) + 0.1 for x in percent_of_total ]\n",
    "        \n",
    "        source = ColumnDataSource(\n",
    "            data = dict(samples=samples, assigned=assigned,\n",
    "                        reads=reads, total_reads=total_reads, viral_reads=viral_reads,\n",
    "                        percent_of_total=percent_of_total, \n",
    "                        percent_of_viral=percent_of_viral,\n",
    "                        contigs=contigs, coverage=coverage,\n",
    "                        colors=colors, alphas=alphas)\n",
    "        )\n",
    "        \n",
    "        TOOLS = \"hover, save, pan, box_zoom, wheel_zoom, reset\"\n",
    "\n",
    "        p = figure(title = title,\n",
    "                  #If desired, the sample can be displayed as \"Run x, sample y\"\n",
    "                  # uncomment the next line if desired\n",
    "                  #x_range = [ \"Run %s, sample %s\" % (x.split('_')[0], x.split('_')[1]) for x in list(sorted(set(samples))) ],\n",
    "                  x_range = list(sorted(set(samples))),\n",
    "                  y_range = list(reversed(sorted(set(assigned)))), #reverse to order 'from top to bottom'\n",
    "                  x_axis_location = \"above\",\n",
    "                  toolbar_location=\"right\",\n",
    "                  tools = TOOLS)\n",
    "\n",
    "        if len(set(assigned)) > 25:\n",
    "            p.plot_height = int(p.plot_height * 1.2)\n",
    "        else:\n",
    "            pass\n",
    "        p.plot_width = len(set(samples)) * 50\n",
    "        #p.plot_height = len(set(assigned)) * 40\n",
    "        p.grid.grid_line_color = None\n",
    "        p.axis.axis_line_color = None\n",
    "        p.axis.major_tick_line_color = None\n",
    "        #p.axis.major_label_text_font_size = \"12pt\"\n",
    "        if len(set(assigned)) > 15:\n",
    "            p.axis.major_label_text_font_size = \"10pt\"\n",
    "        else:\n",
    "            p.axis.major_label_text_font_size = \"12pt\"\n",
    "        p.axis.major_label_standoff = 0\n",
    "        p.xaxis.major_label_orientation = np.pi/4\n",
    "        p.title.text_color = COLOUR[0]\n",
    "        p.title.text_font_size = \"16pt\"\n",
    "        p.title.align = 'right'\n",
    "\n",
    "        p.rect(\"samples\", \"assigned\", 1, 1, source=source,\n",
    "               color=\"colors\", alpha=\"alphas\", line_color=None)\n",
    "\n",
    "        p.select_one(HoverTool).tooltips = [\n",
    "            ('Sample', \"@samples\"),\n",
    "            ('Taxon' , \"@assigned\"),\n",
    "            ('Number of reads', \"@reads\"),\n",
    "            ('Total reads', \"@total_reads (@percent_of_total %)\"),\n",
    "            ('Estimated viral reads', \"@viral_reads (@percent_of_viral %)\"),\n",
    "            ('Number of contigs', \"@contigs\"),\n",
    "            ('Coverage', \"@coverage %\")\n",
    "        ]\n",
    "\n",
    "        #output_file(filename, title=title)\n",
    "        print(\"The heatmap %s has been created and written to: %s\" % (title, filename))\n",
    "        show(p)\n",
    "        return(None)\n",
    "    \n",
    "    #Create an extra column that is the combination of run ID and sample ID:\n",
    "    dataframe[\"sample\"] = dataframe[\"run_id\"].map(str) + '_' + dataframe[\"sample_id\"].map(str)\n",
    "    \n",
    "    #Create heatmaps\n",
    "    #assignments = dataframe.loc[dataframe['Assigned_Discovered'] == 'Assigned']\n",
    "    assignments = dataframe[dataframe.Assigned_Discovered == \"Assigned\"]\n",
    "    create_heatmap(assignments, \"GenomeDetective assignments\", MAP_A)\n",
    "    \n",
    "    #test two methods of selecting rows from a dataframe\n",
    "    #discoveries = dataframe.groupby(\"Assigned_Discovered\").get_group(\"Discovered\")\n",
    "    discoveries = dataframe[dataframe.Assigned_Discovered == \"Discovered\"]\n",
    "    create_heatmap(discoveries, \"GenomeDetective discoveries\", MAP_D)\n",
    "    \n",
    "    create_heatmap(dataframe, \"GenomeDetective assignments+discoveries\", MAP_AD)\n",
    "    \n",
    "    return(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"fa242596-9a57-4f4e-b6e0-bb301cf12577\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id !== undefined) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var element_id = msg.content.text.trim();\n",
       "            Bokeh.index[element_id].model.document.clear();\n",
       "            delete Bokeh.index[element_id];\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"fa242596-9a57-4f4e-b6e0-bb301cf12577\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n",
       "    }\n",
       "    finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.info(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(js_urls, callback) {\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = js_urls.length;\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var s = document.createElement('script');\n",
       "      s.src = url;\n",
       "      s.async = false;\n",
       "      s.onreadystatechange = s.onload = function() {\n",
       "        root._bokeh_is_loading--;\n",
       "        if (root._bokeh_is_loading === 0) {\n",
       "          console.log(\"Bokeh: all BokehJS libraries loaded\");\n",
       "          run_callbacks()\n",
       "        }\n",
       "      };\n",
       "      s.onerror = function() {\n",
       "        console.warn(\"failed to load library \" + url);\n",
       "      };\n",
       "      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "    }\n",
       "  };var element = document.getElementById(\"fa242596-9a57-4f4e-b6e0-bb301cf12577\");\n",
       "  if (element == null) {\n",
       "    console.log(\"Bokeh: ERROR: autoload.js configured with elementid 'fa242596-9a57-4f4e-b6e0-bb301cf12577' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.15.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.15.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.15.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-0.12.15.min.js\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-0.12.15.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.15.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.15.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.15.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.15.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.15.min.css\");\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"fa242596-9a57-4f4e-b6e0-bb301cf12577\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(js_urls, function() {\n",
       "      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"fa242596-9a57-4f4e-b6e0-bb301cf12577\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n    }\n    finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.info(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(js_urls, callback) {\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = js_urls.length;\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var s = document.createElement('script');\n      s.src = url;\n      s.async = false;\n      s.onreadystatechange = s.onload = function() {\n        root._bokeh_is_loading--;\n        if (root._bokeh_is_loading === 0) {\n          console.log(\"Bokeh: all BokehJS libraries loaded\");\n          run_callbacks()\n        }\n      };\n      s.onerror = function() {\n        console.warn(\"failed to load library \" + url);\n      };\n      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.getElementsByTagName(\"head\")[0].appendChild(s);\n    }\n  };var element = document.getElementById(\"fa242596-9a57-4f4e-b6e0-bb301cf12577\");\n  if (element == null) {\n    console.log(\"Bokeh: ERROR: autoload.js configured with elementid 'fa242596-9a57-4f4e-b6e0-bb301cf12577' but no matching script tag was found. \")\n    return false;\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.15.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.15.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.15.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-0.12.15.min.js\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-0.12.15.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.15.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.15.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.15.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.15.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.15.min.css\");\n    }\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"fa242596-9a57-4f4e-b6e0-bb301cf12577\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(js_urls, function() {\n      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The heatmap GenomeDetective assignments has been created and written to: ../tmp/Test_A_heatmap.html\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div class=\"bk-root\">\n",
       "    <div class=\"bk-plotdiv\" id=\"1900db14-a0ec-48c6-80e5-4fbe5ba27303\"></div>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"da8c61ac-627c-4a15-a80b-e3b10b551ec7\":{\"roots\":{\"references\":[{\"attributes\":{},\"id\":\"fffa9a0c-ed42-4f18-9dca-cb2e831b05f5\",\"type\":\"CategoricalScale\"},{\"attributes\":{\"axis_line_color\":{\"value\":null},\"formatter\":{\"id\":\"55114bf8-fe08-4833-bc86-5154e05b48d7\",\"type\":\"CategoricalTickFormatter\"},\"major_label_orientation\":0.7853981633974483,\"major_label_standoff\":0,\"major_label_text_font_size\":{\"value\":\"12pt\"},\"major_tick_line_color\":{\"value\":null},\"plot\":{\"id\":\"6d3e5029-b274-41d3-ae2f-ca8360778a03\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"f7775e44-876a-465c-8a3e-93c699eeb47d\",\"type\":\"CategoricalTicker\"}},\"id\":\"57b25755-15bb-41c0-a751-bd1bece27e2d\",\"type\":\"CategoricalAxis\"},{\"attributes\":{},\"id\":\"aee5659d-b87f-4768-8ad0-efb595441756\",\"type\":\"WheelZoomTool\"},{\"attributes\":{},\"id\":\"abe1d90d-3a81-445c-b9b7-8034aa170d01\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{},\"id\":\"21b50d78-9c32-4295-8dc7-541bc70342dc\",\"type\":\"CategoricalScale\"},{\"attributes\":{\"data_source\":{\"id\":\"0d6455c1-857f-425a-bc31-991c2ed610da\",\"type\":\"ColumnDataSource\"},\"glyph\":{\"id\":\"d08087f5-fba2-46df-9671-643a3df9f7bf\",\"type\":\"Rect\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"c056be3f-8585-4c6f-a8ea-e4a1a21e5a99\",\"type\":\"Rect\"},\"selection_glyph\":null,\"view\":{\"id\":\"eb5586b7-c60d-4238-ba7d-ef1ee0caeafa\",\"type\":\"CDSView\"}},\"id\":\"b59c7eee-6007-4488-9f50-a3e34ffb28b7\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"f7775e44-876a-465c-8a3e-93c699eeb47d\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"axis_line_color\":{\"value\":null},\"formatter\":{\"id\":\"abe1d90d-3a81-445c-b9b7-8034aa170d01\",\"type\":\"CategoricalTickFormatter\"},\"major_label_standoff\":0,\"major_label_text_font_size\":{\"value\":\"12pt\"},\"major_tick_line_color\":{\"value\":null},\"plot\":{\"id\":\"6d3e5029-b274-41d3-ae2f-ca8360778a03\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"83c949e8-97b1-48b6-b8fc-44d701b24d61\",\"type\":\"CategoricalTicker\"}},\"id\":\"d660f5b2-0168-41de-ab2b-9b405eb79adc\",\"type\":\"CategoricalAxis\"},{\"attributes\":{},\"id\":\"55114bf8-fe08-4833-bc86-5154e05b48d7\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"945180f9-ad16-499c-933c-7992e5da0b02\",\"type\":\"HoverTool\"},{\"id\":\"fa5862fb-e759-45f3-9a21-60b2d1e6a7a4\",\"type\":\"SaveTool\"},{\"id\":\"2b7b0423-bb48-4237-ac49-351725bba7ea\",\"type\":\"PanTool\"},{\"id\":\"feab0adf-325d-4d0e-b977-baf1eb9bab08\",\"type\":\"BoxZoomTool\"},{\"id\":\"aee5659d-b87f-4768-8ad0-efb595441756\",\"type\":\"WheelZoomTool\"},{\"id\":\"566d5231-7b89-4d70-b23c-67fceabfd8ac\",\"type\":\"ResetTool\"}]},\"id\":\"e1879c32-ae20-4fc4-86b1-5674ff462846\",\"type\":\"Toolbar\"},{\"attributes\":{\"grid_line_color\":{\"value\":null},\"plot\":{\"id\":\"6d3e5029-b274-41d3-ae2f-ca8360778a03\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"f7775e44-876a-465c-8a3e-93c699eeb47d\",\"type\":\"CategoricalTicker\"}},\"id\":\"7ed4610d-2ccf-47a5-b911-e980cf2e27ba\",\"type\":\"Grid\"},{\"attributes\":{\"above\":[{\"id\":\"57b25755-15bb-41c0-a751-bd1bece27e2d\",\"type\":\"CategoricalAxis\"}],\"left\":[{\"id\":\"d660f5b2-0168-41de-ab2b-9b405eb79adc\",\"type\":\"CategoricalAxis\"}],\"plot_width\":200,\"renderers\":[{\"id\":\"57b25755-15bb-41c0-a751-bd1bece27e2d\",\"type\":\"CategoricalAxis\"},{\"id\":\"7ed4610d-2ccf-47a5-b911-e980cf2e27ba\",\"type\":\"Grid\"},{\"id\":\"d660f5b2-0168-41de-ab2b-9b405eb79adc\",\"type\":\"CategoricalAxis\"},{\"id\":\"4238b2ca-93e9-42ef-93cb-5890cf7cd1c8\",\"type\":\"Grid\"},{\"id\":\"a711403f-76ea-4498-9369-26d4d410b238\",\"type\":\"BoxAnnotation\"},{\"id\":\"b59c7eee-6007-4488-9f50-a3e34ffb28b7\",\"type\":\"GlyphRenderer\"}],\"title\":{\"id\":\"e9603b8d-08cb-4772-b087-0ff8bc271167\",\"type\":\"Title\"},\"toolbar\":{\"id\":\"e1879c32-ae20-4fc4-86b1-5674ff462846\",\"type\":\"Toolbar\"},\"x_range\":{\"id\":\"aa8a860d-0526-424e-bd7e-93656f86ddf0\",\"type\":\"FactorRange\"},\"x_scale\":{\"id\":\"fffa9a0c-ed42-4f18-9dca-cb2e831b05f5\",\"type\":\"CategoricalScale\"},\"y_range\":{\"id\":\"330c5fc5-6704-4d2b-be8c-a6cac5ef9cac\",\"type\":\"FactorRange\"},\"y_scale\":{\"id\":\"21b50d78-9c32-4295-8dc7-541bc70342dc\",\"type\":\"CategoricalScale\"}},\"id\":\"6d3e5029-b274-41d3-ae2f-ca8360778a03\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{},\"id\":\"83c949e8-97b1-48b6-b8fc-44d701b24d61\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"height\":{\"units\":\"data\",\"value\":1},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"width\":{\"units\":\"data\",\"value\":1},\"x\":{\"field\":\"samples\"},\"y\":{\"field\":\"assigned\"}},\"id\":\"c056be3f-8585-4c6f-a8ea-e4a1a21e5a99\",\"type\":\"Rect\"},{\"attributes\":{\"callback\":null,\"factors\":[\"3_1\",\"4_D\",\"5_5061600092_S2\",\"6_A1\"]},\"id\":\"aa8a860d-0526-424e-bd7e-93656f86ddf0\",\"type\":\"FactorRange\"},{\"attributes\":{\"fill_alpha\":{\"field\":\"alphas\"},\"fill_color\":{\"field\":\"colors\"},\"height\":{\"units\":\"data\",\"value\":1},\"line_alpha\":{\"field\":\"alphas\"},\"line_color\":{\"value\":null},\"width\":{\"units\":\"data\",\"value\":1},\"x\":{\"field\":\"samples\"},\"y\":{\"field\":\"assigned\"}},\"id\":\"d08087f5-fba2-46df-9671-643a3df9f7bf\",\"type\":\"Rect\"},{\"attributes\":{\"callback\":null,\"factors\":[\"uncultured crAssphage\",\"Pepper mild mottle virus\",\"Parechovirus A\",\"Norovirus\",\"Escherichia virus phiV10\",\"Equartevirus\",\"Enterovirus B\",\"Enterovirus A\"]},\"id\":\"330c5fc5-6704-4d2b-be8c-a6cac5ef9cac\",\"type\":\"FactorRange\"},{\"attributes\":{\"dimension\":1,\"grid_line_color\":{\"value\":null},\"plot\":{\"id\":\"6d3e5029-b274-41d3-ae2f-ca8360778a03\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"83c949e8-97b1-48b6-b8fc-44d701b24d61\",\"type\":\"CategoricalTicker\"}},\"id\":\"4238b2ca-93e9-42ef-93cb-5890cf7cd1c8\",\"type\":\"Grid\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":{\"value\":0.5},\"fill_color\":{\"value\":\"lightgrey\"},\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":{\"value\":1.0},\"line_color\":{\"value\":\"black\"},\"line_dash\":[4,4],\"line_width\":{\"value\":2},\"plot\":null,\"render_mode\":\"css\",\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"a711403f-76ea-4498-9369-26d4d410b238\",\"type\":\"BoxAnnotation\"},{\"attributes\":{\"callback\":null,\"tooltips\":[[\"Sample\",\"@samples\"],[\"Taxon\",\"@assigned\"],[\"Number of reads\",\"@reads\"],[\"Total reads\",\"@total_reads (@percent_of_total %)\"],[\"Estimated viral reads\",\"@viral_reads (@percent_of_viral %)\"],[\"Number of contigs\",\"@contigs\"],[\"Coverage\",\"@coverage %\"]]},\"id\":\"945180f9-ad16-499c-933c-7992e5da0b02\",\"type\":\"HoverTool\"},{\"attributes\":{\"source\":{\"id\":\"0d6455c1-857f-425a-bc31-991c2ed610da\",\"type\":\"ColumnDataSource\"}},\"id\":\"eb5586b7-c60d-4238-ba7d-ef1ee0caeafa\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"566d5231-7b89-4d70-b23c-67fceabfd8ac\",\"type\":\"ResetTool\"},{\"attributes\":{\"callback\":null,\"column_names\":[\"samples\",\"assigned\",\"reads\",\"total_reads\",\"viral_reads\",\"percent_of_total\",\"percent_of_viral\",\"contigs\",\"coverage\",\"colors\",\"alphas\"],\"data\":{\"alphas\":[1.0,0.24550343536211372,0.10995474245858046,0.1002820600191296,0.5420005539788172,0.26404874356631064,0.10803590374165291,0.10022139117077532,1.0,0.30040662529562445,0.17317489760468974,0.10019144527434412,0.10087473369253164],\"assigned\":[\"Norovirus\",\"uncultured crAssphage\",\"Norovirus\",\"Pepper mild mottle virus\",\"Norovirus\",\"Parechovirus A\",\"Equartevirus\",\"Enterovirus B\",\"Enterovirus A\",\"Norovirus\",\"Parechovirus A\",\"Enterovirus B\",\"Escherichia virus phiV10\"],\"colors\":[\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\"],\"contigs\":{\"__ndarray__\":\"AAAAAAAA8D8AAAAAAAAoQAAAAAAAAPA/AAAAAAAACEAAAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAAhAAAAAAAAAFEA=\",\"dtype\":\"float64\",\"shape\":[13]},\"coverage\":{\"__ndarray__\":\"+KX15+70WEADrNwaX4RTQKjqhA+IOllAShMddt+HUkASdYc1cv1YQAIRXvGso1hA1A7NSrzcWEDdJXdFp3RWQFPaVlkx31hAXzHqISctWUD53sD4g+FYQGw1Qp2JrE5AbvPJyjTeS0A=\",\"dtype\":\"float64\",\"shape\":[13]},\"percent_of_total\":{\"__ndarray__\":\"rFsA4pbxT0DdPz4qeJciQFrU6pD4WeQ/dBKYgdtzkj8cYASC/zw8QCpokNIX9iRAhdoIcLdt4D9J40VRlveMP9XbRDltJE5AnN17U1+bKUCqTOt0J7MSQPgo6oaJDIk/VLyr2+OcrD8=\",\"dtype\":\"float64\",\"shape\":[13]},\"percent_of_viral\":{\"__ndarray__\":\"addLqABBV0DIxHlmdRErQEhXPzTLhxlAfcRknvklxz+oouSsWY5VQPJKaQFHAEBA8Z6pQPoU+T8zrjkYyRymP27seLJjFlZAUrv9o5nDMkA0LN6Ey2cbQPgZbjXvWpI/ozaJY3v3tD8=\",\"dtype\":\"float64\",\"shape\":[13]},\"reads\":[542592,78949,5541,157,4599864,1707242,83629,2304,1212499,257515,94027,246,1124],\"samples\":[\"3_1\",\"3_1\",\"4_D\",\"4_D\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\"],\"total_reads\":[849294,849294,871250,871250,16289466,16289466,16289466,16289466,2011292,2011292,2011292,2011292,2011292],\"viral_reads\":[583334,583334,86814,86814,5334770,5334770,5334770,5334770,1372384,1372384,1372384,1372384,1372384]},\"selected\":null,\"selection_policy\":null},\"id\":\"0d6455c1-857f-425a-bc31-991c2ed610da\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"overlay\":{\"id\":\"a711403f-76ea-4498-9369-26d4d410b238\",\"type\":\"BoxAnnotation\"}},\"id\":\"feab0adf-325d-4d0e-b977-baf1eb9bab08\",\"type\":\"BoxZoomTool\"},{\"attributes\":{},\"id\":\"fa5862fb-e759-45f3-9a21-60b2d1e6a7a4\",\"type\":\"SaveTool\"},{\"attributes\":{\"align\":\"right\",\"plot\":null,\"text\":\"GenomeDetective assignments\",\"text_color\":{\"field\":\"-\"},\"text_font_size\":{\"value\":\"16pt\"}},\"id\":\"e9603b8d-08cb-4772-b087-0ff8bc271167\",\"type\":\"Title\"},{\"attributes\":{},\"id\":\"2b7b0423-bb48-4237-ac49-351725bba7ea\",\"type\":\"PanTool\"}],\"root_ids\":[\"6d3e5029-b274-41d3-ae2f-ca8360778a03\"]},\"title\":\"Bokeh Application\",\"version\":\"0.12.15\"}};\n",
       "  var render_items = [{\"docid\":\"da8c61ac-627c-4a15-a80b-e3b10b551ec7\",\"elementid\":\"1900db14-a0ec-48c6-80e5-4fbe5ba27303\",\"modelid\":\"6d3e5029-b274-41d3-ae2f-ca8360778a03\"}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        embed_document(root);\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "      attempts++;\n",
       "      if (attempts > 100) {\n",
       "        console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\")\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "6d3e5029-b274-41d3-ae2f-ca8360778a03"
      }
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The heatmap GenomeDetective discoveries has been created and written to: ../tmp/Test_D_heatmap.html\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div class=\"bk-root\">\n",
       "    <div class=\"bk-plotdiv\" id=\"38c9f8dc-6d0d-454a-bd38-409a570333b4\"></div>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"b5e6cd3c-2893-47a8-b742-57ee51f30c70\":{\"roots\":{\"references\":[{\"attributes\":{\"fill_alpha\":{\"field\":\"alphas\"},\"fill_color\":{\"field\":\"colors\"},\"height\":{\"units\":\"data\",\"value\":1},\"line_alpha\":{\"field\":\"alphas\"},\"line_color\":{\"value\":null},\"width\":{\"units\":\"data\",\"value\":1},\"x\":{\"field\":\"samples\"},\"y\":{\"field\":\"assigned\"}},\"id\":\"f6c5e51c-58ef-4332-b482-10d8f68db2e2\",\"type\":\"Rect\"},{\"attributes\":{},\"id\":\"f9ad6c49-629d-45f9-bcfa-b54528a0954a\",\"type\":\"PanTool\"},{\"attributes\":{\"overlay\":{\"id\":\"e66e1ea3-0e12-4908-8687-d69796d01872\",\"type\":\"BoxAnnotation\"}},\"id\":\"ea5d5349-c384-4255-a401-800908a192b3\",\"type\":\"BoxZoomTool\"},{\"attributes\":{},\"id\":\"5308b3f0-e4bd-4261-86e6-e49eb4ca4b16\",\"type\":\"SaveTool\"},{\"attributes\":{},\"id\":\"e96a19d4-16e0-4b37-a0ac-f7922a11a270\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{\"callback\":null,\"factors\":[\"3_1\",\"4_D\",\"5_5061600092_S2\",\"6_A1\"]},\"id\":\"fc8eb07c-0dfa-4001-a2bc-d730d151d99f\",\"type\":\"FactorRange\"},{\"attributes\":{},\"id\":\"51b62810-4335-4b2e-b019-f8e219620847\",\"type\":\"ResetTool\"},{\"attributes\":{\"align\":\"right\",\"plot\":null,\"text\":\"GenomeDetective discoveries\",\"text_color\":{\"field\":\"-\"},\"text_font_size\":{\"value\":\"16pt\"}},\"id\":\"a56dda45-7a4b-4007-822f-4c74f9e32573\",\"type\":\"Title\"},{\"attributes\":{\"callback\":null,\"tooltips\":[[\"Sample\",\"@samples\"],[\"Taxon\",\"@assigned\"],[\"Number of reads\",\"@reads\"],[\"Total reads\",\"@total_reads (@percent_of_total %)\"],[\"Estimated viral reads\",\"@viral_reads (@percent_of_viral %)\"],[\"Number of contigs\",\"@contigs\"],[\"Coverage\",\"@coverage %\"]]},\"id\":\"973ea13e-0568-4265-b4d6-781466447ab2\",\"type\":\"HoverTool\"},{\"attributes\":{\"callback\":null,\"column_names\":[\"samples\",\"assigned\",\"reads\",\"total_reads\",\"viral_reads\",\"percent_of_total\",\"percent_of_viral\",\"contigs\",\"coverage\",\"colors\",\"alphas\"],\"data\":{\"alphas\":[0.3122690963516434,0.2551108597661846,0.17029541192969733,0.1626896788356645,0.12489149012592561,0.11912957111529468,0.11428955914636471,0.1110628845004114,0.1110628845004114,0.10829716337530854,0.10829716337530854,0.10414858168765427,0.10253524436467762,0.10253524436467762,0.10207429084382715,0.1011523838021262,0.10046095352085048,0.7342394967423052,1.0,0.4383509323747472,0.3066951246910807,0.13370029206919792,0.13033026286227814,0.11123343068973265,0.10134801168276793,0.10044933722758931,0.10140593166055445,0.12326396320370434,0.11353059016909664,0.10106946938281493,0.14700642267668226,0.14155640265619737,0.13065636261522756,0.11917239185777724,0.10973217860800875,0.10194643572160175,0.1007785742886407],\"assigned\":[\"Lactococcus virus bIL67\",\"Sk1virus\",\"Bacteroides phage B124-14\",\"Lactococcus virus c2\",\"Croceibacter phage P2559Y\",\"Parabacteroides phage YZ-2015a\",\"Croceibacter phage P2559S\",\"Escherichia virus Cajan\",\"Escherichia virus Seurat\",\"Bacillus phage VMY22\",\"Chimpanzee faeces associated microphage 2\",\"Streptococcus phage 315.1\",\"Sk1virus\",\"Escherichia phage slur01\",\"Bacillus virus phi29\",\"Clostridium phage vB_CpeS-CP51\",\"Escherichia virus YD2008s\",\"Otarine picobirnavirus (segment 1)\",\"Sk1virus\",\"Bacillus virus B103\",\"Bacillus phage MG-B1\",\"Actinomyces virus Av1\",\"Lactococcus phage Tuc2009\",\"Lactococcus phage bIL286\",\"Leuconostoc phage Lmd1\",\"Lactobacillus phage JCL1032\",\"Alphatorquevirus\",\"P1virus\",\"Escherichia virus HK022\",\"Salmonella phage SJ46\",\"Bacillus virus GA1\",\"Escherichia phage TL-2011b\",\"Clostridium phage phiMMP04\",\"Actinomyces virus Av1\",\"Bacillus virus phi29\",\"Escherichia virus 191\",\"Escherichia phage P13374\"],\"colors\":[\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\"],\"contigs\":{\"__ndarray__\":\"AAAAAAAAHEAAAAAAAAAmQAAAAAAAAAhAAAAAAAAACEAAAAAAAAAQQAAAAAAAAABAAAAAAAAA8D8AAAAAAAAkQAAAAAAAABhAAAAAAAAAAEAAAAAAAAAAQAAAAAAAAABAAAAAAAAAAEAAAAAAAAAQQAAAAAAAAABAAAAAAAAAAEAAAAAAAADwPwAAAAAAAPA/AAAAAAAAKEAAAAAAAADwPwAAAAAAAABAAAAAAAAAGEAAAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAAAUQAAAAAAAAPA/AAAAAAAAAEAAAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8=\",\"dtype\":\"float64\",\"shape\":[37]},\"coverage\":{\"__ndarray__\":\"lwX5rkr3TEBy5w1Al1FNQH8p/NSrdiVAhYVDppCcREBDncMS1QMiQAEy93B4b0NAv4YgRQ3XEEABqAxX/WEcQHXJJ7kLExpAFdAKTgf9F0Bd88MiZdBEQCQ/lIg1qghAXGb6H1JiBECeJw/NAhYFQBHzDoUlgxVAUxiPV/A19D8WF8JguqDuP5aBFMyTn0VAnmv88jWTUUAb9ncxqAgjQIA7eB3ZQh5AqJ2r8zgcMUDF6cVWblAQQIRd6/1BMAVANdLmeLqr+D/qdWdmEvfrP1X0GnuCByBAweELKQI29j+SgFK9acD1P9mg6Q/a2co/QwAaLkvrJEB8uSxZ3+oyQPfvX4UJzxBAlp0y79wPIEAetWPNIS8JQFQCv1T5IcQ/8WzNcOgtxD8=\",\"dtype\":\"float64\",\"shape\":[37]},\"percent_of_total\":{\"__ndarray__\":\"2Wri2OvCuz99ILb8OEm0P5JsXNcUY6I/p9QWZ8ploD/rfmBJFAuKP4B08hvGA4Q/bSmxcMvmfT8K/+PrSiZ3Pwr/4+tKJnc/SP/qMLhccT9I/+owuFxxP0j/6jC4XGE/yak7WG84VT/JqTtYbzhVP0j/6jC4XFE/iFSTxJNKQz8NVIU6ud0uP72UNLWyvNQ/WVBz5xZZ4D8gFj6eHSDGPwkb0PtMCLs/oRFQVj2hkT+7H5DO1LuPP4IXwB2ngXc/kRYAWvKQRj8XHgB4mBYuP6zICCIqiUc/NZ2DMidXiD+ZFuKzOFB8P+v4rK4+50E/XhKeFTyXmD9IaTY4Wr2VPxwXZ32WCZA/EsF0Qz4PhD95JVLkcV10P8fqdIPBSlA/pXchnzUROj8=\",\"dtype\":\"float64\",\"shape\":[37]},\"percent_of_viral\":{\"__ndarray__\":\"GCgw2pc1xD8Mm00q+Yi9P2y5hHAvxao/wjrvcLHfpz9PVBhDZ/WSP/bD51jcI40/SGolFG3EhT/Vg07mItqAP9WDTuYi2oA/v8V1WTRHeT+/xXVZNEd5P7/FdVk0R2k/sfGP+z/lXj+x8Y/7P+VeP7/FdVk0R1k/uNuCKjoWTD/Gr2iILng2P04s2A6lAwpAahWfPhqCFEAwK8shgsH7P8gdWrCu9PA/X1O4dsIdxj/VF9lqlefDP37E9UgDfa0/EvrCZA1PfD+3pixDXt9iP4ndrmNj92E/mbA+3aGUoj+20nRVBp2VPzgwgCFcVVs/OaxUcfoEoj/5kBfMR9yfPwcC9J7tgJc/2xZdg+9ljT+E/n4zi9h9P2mYZVxv4Fc/ukZRsPIZQz8=\",\"dtype\":\"float64\",\"shape\":[37]},\"reads\":[921,673,305,272,108,83,62,48,48,36,36,18,11,11,9,5,2,2823,4451,1506,920,150,135,50,6,2,117,1936,1126,89,483,427,315,197,100,20,8],\"samples\":[\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\"],\"total_reads\":[849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,871250,871250,871250,871250,871250,871250,871250,871250,871250,16289466,16289466,16289466,16289466,2011292,2011292,2011292,2011292,2011292,2011292,2011292],\"viral_reads\":[583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,86814,86814,86814,86814,86814,86814,86814,86814,86814,5334770,5334770,5334770,5334770,1372384,1372384,1372384,1372384,1372384,1372384,1372384]},\"selected\":null,\"selection_policy\":null},\"id\":\"478eefb0-e5c2-4447-afbf-a4ae8067f884\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":{\"value\":0.5},\"fill_color\":{\"value\":\"lightgrey\"},\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":{\"value\":1.0},\"line_color\":{\"value\":\"black\"},\"line_dash\":[4,4],\"line_width\":{\"value\":2},\"plot\":null,\"render_mode\":\"css\",\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"e66e1ea3-0e12-4908-8687-d69796d01872\",\"type\":\"BoxAnnotation\"},{\"attributes\":{},\"id\":\"fb818327-c241-49ba-a105-b8112979d50d\",\"type\":\"CategoricalScale\"},{\"attributes\":{\"source\":{\"id\":\"478eefb0-e5c2-4447-afbf-a4ae8067f884\",\"type\":\"ColumnDataSource\"}},\"id\":\"c2207ecd-d17d-4ac0-8a5a-6f7eb1944753\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"91b2013e-c56b-43fe-a276-ed86a645f5e5\",\"type\":\"CategoricalScale\"},{\"attributes\":{},\"id\":\"468f429f-5678-4795-a4f0-4517f07e89a1\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{\"axis_line_color\":{\"value\":null},\"formatter\":{\"id\":\"e96a19d4-16e0-4b37-a0ac-f7922a11a270\",\"type\":\"CategoricalTickFormatter\"},\"major_label_standoff\":0,\"major_label_text_font_size\":{\"value\":\"10pt\"},\"major_tick_line_color\":{\"value\":null},\"plot\":{\"id\":\"b6cea7c1-3d73-4bca-9957-3e0468a8c024\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"7aab3cd9-5782-4d2a-8fe9-d63a89a68d2a\",\"type\":\"CategoricalTicker\"}},\"id\":\"a0d3aaf4-3b70-46d6-bd1e-e17bb3fe159b\",\"type\":\"CategoricalAxis\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"973ea13e-0568-4265-b4d6-781466447ab2\",\"type\":\"HoverTool\"},{\"id\":\"5308b3f0-e4bd-4261-86e6-e49eb4ca4b16\",\"type\":\"SaveTool\"},{\"id\":\"f9ad6c49-629d-45f9-bcfa-b54528a0954a\",\"type\":\"PanTool\"},{\"id\":\"ea5d5349-c384-4255-a401-800908a192b3\",\"type\":\"BoxZoomTool\"},{\"id\":\"21e20f54-3d1b-4ca7-8c64-2ce12a01c2af\",\"type\":\"WheelZoomTool\"},{\"id\":\"51b62810-4335-4b2e-b019-f8e219620847\",\"type\":\"ResetTool\"}]},\"id\":\"ec31219e-2834-4ada-8fb5-a92de6767144\",\"type\":\"Toolbar\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"height\":{\"units\":\"data\",\"value\":1},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"width\":{\"units\":\"data\",\"value\":1},\"x\":{\"field\":\"samples\"},\"y\":{\"field\":\"assigned\"}},\"id\":\"69ef3a55-30b3-4914-a6a6-44600e5f5a45\",\"type\":\"Rect\"},{\"attributes\":{\"dimension\":1,\"grid_line_color\":{\"value\":null},\"plot\":{\"id\":\"b6cea7c1-3d73-4bca-9957-3e0468a8c024\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"7aab3cd9-5782-4d2a-8fe9-d63a89a68d2a\",\"type\":\"CategoricalTicker\"}},\"id\":\"5b49fe87-adc2-4c74-a0b7-e4b96945c0b5\",\"type\":\"Grid\"},{\"attributes\":{\"callback\":null,\"factors\":[\"Streptococcus phage 315.1\",\"Sk1virus\",\"Salmonella phage SJ46\",\"Parabacteroides phage YZ-2015a\",\"P1virus\",\"Otarine picobirnavirus (segment 1)\",\"Leuconostoc phage Lmd1\",\"Lactococcus virus c2\",\"Lactococcus virus bIL67\",\"Lactococcus phage bIL286\",\"Lactococcus phage Tuc2009\",\"Lactobacillus phage JCL1032\",\"Escherichia virus YD2008s\",\"Escherichia virus Seurat\",\"Escherichia virus HK022\",\"Escherichia virus Cajan\",\"Escherichia virus 191\",\"Escherichia phage slur01\",\"Escherichia phage TL-2011b\",\"Escherichia phage P13374\",\"Croceibacter phage P2559Y\",\"Croceibacter phage P2559S\",\"Clostridium phage vB_CpeS-CP51\",\"Clostridium phage phiMMP04\",\"Chimpanzee faeces associated microphage 2\",\"Bacteroides phage B124-14\",\"Bacillus virus phi29\",\"Bacillus virus GA1\",\"Bacillus virus B103\",\"Bacillus phage VMY22\",\"Bacillus phage MG-B1\",\"Alphatorquevirus\",\"Actinomyces virus Av1\"]},\"id\":\"2c90d602-c3a4-46a1-9a6c-a1861a6af673\",\"type\":\"FactorRange\"},{\"attributes\":{},\"id\":\"7aab3cd9-5782-4d2a-8fe9-d63a89a68d2a\",\"type\":\"CategoricalTicker\"},{\"attributes\":{},\"id\":\"e9b6acf5-68db-47f5-aa71-7e3e67ca1f7b\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"axis_line_color\":{\"value\":null},\"formatter\":{\"id\":\"468f429f-5678-4795-a4f0-4517f07e89a1\",\"type\":\"CategoricalTickFormatter\"},\"major_label_orientation\":0.7853981633974483,\"major_label_standoff\":0,\"major_label_text_font_size\":{\"value\":\"10pt\"},\"major_tick_line_color\":{\"value\":null},\"plot\":{\"id\":\"b6cea7c1-3d73-4bca-9957-3e0468a8c024\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"e9b6acf5-68db-47f5-aa71-7e3e67ca1f7b\",\"type\":\"CategoricalTicker\"}},\"id\":\"0fa4bebb-05ce-44ab-a4d7-2bf52448ef31\",\"type\":\"CategoricalAxis\"},{\"attributes\":{\"above\":[{\"id\":\"0fa4bebb-05ce-44ab-a4d7-2bf52448ef31\",\"type\":\"CategoricalAxis\"}],\"left\":[{\"id\":\"a0d3aaf4-3b70-46d6-bd1e-e17bb3fe159b\",\"type\":\"CategoricalAxis\"}],\"plot_height\":720,\"plot_width\":200,\"renderers\":[{\"id\":\"0fa4bebb-05ce-44ab-a4d7-2bf52448ef31\",\"type\":\"CategoricalAxis\"},{\"id\":\"4a591ef8-2c6a-4401-b1bb-b707c524027d\",\"type\":\"Grid\"},{\"id\":\"a0d3aaf4-3b70-46d6-bd1e-e17bb3fe159b\",\"type\":\"CategoricalAxis\"},{\"id\":\"5b49fe87-adc2-4c74-a0b7-e4b96945c0b5\",\"type\":\"Grid\"},{\"id\":\"e66e1ea3-0e12-4908-8687-d69796d01872\",\"type\":\"BoxAnnotation\"},{\"id\":\"92ceeab1-ddb8-4a4e-bc61-444b0c90d906\",\"type\":\"GlyphRenderer\"}],\"title\":{\"id\":\"a56dda45-7a4b-4007-822f-4c74f9e32573\",\"type\":\"Title\"},\"toolbar\":{\"id\":\"ec31219e-2834-4ada-8fb5-a92de6767144\",\"type\":\"Toolbar\"},\"x_range\":{\"id\":\"fc8eb07c-0dfa-4001-a2bc-d730d151d99f\",\"type\":\"FactorRange\"},\"x_scale\":{\"id\":\"91b2013e-c56b-43fe-a276-ed86a645f5e5\",\"type\":\"CategoricalScale\"},\"y_range\":{\"id\":\"2c90d602-c3a4-46a1-9a6c-a1861a6af673\",\"type\":\"FactorRange\"},\"y_scale\":{\"id\":\"fb818327-c241-49ba-a105-b8112979d50d\",\"type\":\"CategoricalScale\"}},\"id\":\"b6cea7c1-3d73-4bca-9957-3e0468a8c024\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"grid_line_color\":{\"value\":null},\"plot\":{\"id\":\"b6cea7c1-3d73-4bca-9957-3e0468a8c024\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"e9b6acf5-68db-47f5-aa71-7e3e67ca1f7b\",\"type\":\"CategoricalTicker\"}},\"id\":\"4a591ef8-2c6a-4401-b1bb-b707c524027d\",\"type\":\"Grid\"},{\"attributes\":{\"data_source\":{\"id\":\"478eefb0-e5c2-4447-afbf-a4ae8067f884\",\"type\":\"ColumnDataSource\"},\"glyph\":{\"id\":\"f6c5e51c-58ef-4332-b482-10d8f68db2e2\",\"type\":\"Rect\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"69ef3a55-30b3-4914-a6a6-44600e5f5a45\",\"type\":\"Rect\"},\"selection_glyph\":null,\"view\":{\"id\":\"c2207ecd-d17d-4ac0-8a5a-6f7eb1944753\",\"type\":\"CDSView\"}},\"id\":\"92ceeab1-ddb8-4a4e-bc61-444b0c90d906\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"21e20f54-3d1b-4ca7-8c64-2ce12a01c2af\",\"type\":\"WheelZoomTool\"}],\"root_ids\":[\"b6cea7c1-3d73-4bca-9957-3e0468a8c024\"]},\"title\":\"Bokeh Application\",\"version\":\"0.12.15\"}};\n",
       "  var render_items = [{\"docid\":\"b5e6cd3c-2893-47a8-b742-57ee51f30c70\",\"elementid\":\"38c9f8dc-6d0d-454a-bd38-409a570333b4\",\"modelid\":\"b6cea7c1-3d73-4bca-9957-3e0468a8c024\"}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        embed_document(root);\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "      attempts++;\n",
       "      if (attempts > 100) {\n",
       "        console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\")\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "b6cea7c1-3d73-4bca-9957-3e0468a8c024"
      }
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The heatmap GenomeDetective assignments+discoveries has been created and written to: ../tmp/Test_AD_heatmap.html\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div class=\"bk-root\">\n",
       "    <div class=\"bk-plotdiv\" id=\"5b4764f8-0b6f-49d8-ab26-4cbf13a6fa01\"></div>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"03941f5a-edb1-4dc9-995c-a7d31b56987b\":{\"roots\":{\"references\":[{\"attributes\":{\"axis_line_color\":{\"value\":null},\"formatter\":{\"id\":\"1fa7cd6e-297c-42b2-b84b-9def0b534dc4\",\"type\":\"CategoricalTickFormatter\"},\"major_label_standoff\":0,\"major_label_text_font_size\":{\"value\":\"10pt\"},\"major_tick_line_color\":{\"value\":null},\"plot\":{\"id\":\"ec6bf758-84e2-4a10-b53e-2f64a6ffd6fd\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"166d91ce-8ba1-4c2d-9a84-f6ace1755c13\",\"type\":\"CategoricalTicker\"}},\"id\":\"1fd8bda6-c05f-4357-b404-1edbeaa36f52\",\"type\":\"CategoricalAxis\"},{\"attributes\":{},\"id\":\"3a6812b9-905b-41e5-9c61-25c03fd983ed\",\"type\":\"PanTool\"},{\"attributes\":{\"callback\":null,\"tooltips\":[[\"Sample\",\"@samples\"],[\"Taxon\",\"@assigned\"],[\"Number of reads\",\"@reads\"],[\"Total reads\",\"@total_reads (@percent_of_total %)\"],[\"Estimated viral reads\",\"@viral_reads (@percent_of_viral %)\"],[\"Number of contigs\",\"@contigs\"],[\"Coverage\",\"@coverage %\"]]},\"id\":\"203ebb8f-3b65-4ed8-9c05-50f4c106cd0d\",\"type\":\"HoverTool\"},{\"attributes\":{},\"id\":\"166d91ce-8ba1-4c2d-9a84-f6ace1755c13\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"dimension\":1,\"grid_line_color\":{\"value\":null},\"plot\":{\"id\":\"ec6bf758-84e2-4a10-b53e-2f64a6ffd6fd\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"166d91ce-8ba1-4c2d-9a84-f6ace1755c13\",\"type\":\"CategoricalTicker\"}},\"id\":\"7ee9875a-715a-4b4c-83fd-77f0aa6f93ef\",\"type\":\"Grid\"},{\"attributes\":{\"overlay\":{\"id\":\"af50bcbe-ec5e-42eb-a370-5713d8b11bee\",\"type\":\"BoxAnnotation\"}},\"id\":\"94eed387-0fd8-4070-92e9-6d9147332354\",\"type\":\"BoxZoomTool\"},{\"attributes\":{\"callback\":null,\"column_names\":[\"samples\",\"assigned\",\"reads\",\"total_reads\",\"viral_reads\",\"percent_of_total\",\"percent_of_viral\",\"contigs\",\"coverage\",\"colors\",\"alphas\"],\"data\":{\"alphas\":[1.0,0.24550343536211372,0.10169740799716914,0.10124034265156877,0.10056211665487144,0.10050129747581978,0.10019904458598727,0.10015296945034206,0.1001142663364001,0.10008846426043878,0.10008846426043878,0.1000663481953291,0.1000663481953291,0.10003317409766455,0.10002027305968389,0.10002027305968389,0.10001658704883228,0.10000921502712905,0.10000368601085162,0.10995474245858046,0.1002820600191296,0.10507169066243867,0.10799649137035583,0.10270562031088652,0.10165283578088685,0.10026948409470982,0.10024253568523883,0.10008982803156995,0.1000107793637884,0.10000359312126281,0.5420005539788172,0.26404874356631064,0.10803590374165291,0.10022139117077532,0.10001124252039094,0.1001860300809987,0.10010819724752301,0.10000855200269054,1.0,0.30040662529562445,0.17317489760468974,0.10019144527434412,0.10087473369253164,0.10037588645328539,0.10033230541522332,0.10024514333909917,0.1001533118660398,0.1000778232822537,0.10001556465645074,0.1000062258625803],\"assigned\":[\"Norovirus\",\"uncultured crAssphage\",\"Lactococcus virus bIL67\",\"Sk1virus\",\"Bacteroides phage B124-14\",\"Lactococcus virus c2\",\"Croceibacter phage P2559Y\",\"Parabacteroides phage YZ-2015a\",\"Croceibacter phage P2559S\",\"Escherichia virus Cajan\",\"Escherichia virus Seurat\",\"Bacillus phage VMY22\",\"Chimpanzee faeces associated microphage 2\",\"Streptococcus phage 315.1\",\"Sk1virus\",\"Escherichia phage slur01\",\"Bacillus virus phi29\",\"Clostridium phage vB_CpeS-CP51\",\"Escherichia virus YD2008s\",\"Norovirus\",\"Pepper mild mottle virus\",\"Otarine picobirnavirus (segment 1)\",\"Sk1virus\",\"Bacillus virus B103\",\"Bacillus phage MG-B1\",\"Actinomyces virus Av1\",\"Lactococcus phage Tuc2009\",\"Lactococcus phage bIL286\",\"Leuconostoc phage Lmd1\",\"Lactobacillus phage JCL1032\",\"Norovirus\",\"Parechovirus A\",\"Equartevirus\",\"Enterovirus B\",\"Alphatorquevirus\",\"P1virus\",\"Escherichia virus HK022\",\"Salmonella phage SJ46\",\"Enterovirus A\",\"Norovirus\",\"Parechovirus A\",\"Enterovirus B\",\"Escherichia virus phiV10\",\"Bacillus virus GA1\",\"Escherichia phage TL-2011b\",\"Clostridium phage phiMMP04\",\"Actinomyces virus Av1\",\"Bacillus virus phi29\",\"Escherichia virus 191\",\"Escherichia phage P13374\"],\"colors\":[\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\",\"-\",\"f\"],\"contigs\":{\"__ndarray__\":\"AAAAAAAA8D8AAAAAAAAoQAAAAAAAABxAAAAAAAAAJkAAAAAAAAAIQAAAAAAAAAhAAAAAAAAAEEAAAAAAAAAAQAAAAAAAAPA/AAAAAAAAJEAAAAAAAAAYQAAAAAAAAABAAAAAAAAAAEAAAAAAAAAAQAAAAAAAAABAAAAAAAAAEEAAAAAAAAAAQAAAAAAAAABAAAAAAAAA8D8AAAAAAADwPwAAAAAAAAhAAAAAAAAA8D8AAAAAAAAoQAAAAAAAAPA/AAAAAAAAAEAAAAAAAAAYQAAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPwAAAAAAAAhAAAAAAAAAFEAAAAAAAADwPwAAAAAAABRAAAAAAAAA8D8AAAAAAAAAQAAAAAAAAPA/AAAAAAAA8D8AAAAAAADwPw==\",\"dtype\":\"float64\",\"shape\":[50]},\"coverage\":{\"__ndarray__\":\"+KX15+70WEADrNwaX4RTQJcF+a5K90xAcucNQJdRTUB/KfzUq3YlQIWFQ6aQnERAQ53DEtUDIkABMvdweG9DQL+GIEUN1xBAAagMV/1hHEB1ySe5CxMaQBXQCk4H/RdAXfPDImXQREAkP5SINaoIQFxm+h9SYgRAnicPzQIWBUAR8w6FJYMVQFMYj1fwNfQ/FhfCYLqg7j+o6oQPiDpZQEoTHXbfh1JAloEUzJOfRUCea/zyNZNRQBv2dzGoCCNAgDt4HdlCHkConavzOBwxQMXpxVZuUBBAhF3r/UEwBUA10uZ4uqv4P+p1Z2YS9+s/EnWHNXL9WEACEV7xrKNYQNQOzUq83FhA3SV3Rad0VkBV9Bp7ggcgQMHhCykCNvY/koBSvWnA9T/ZoOkP2tnKP1PaVlkx31hAXzHqISctWUD53sD4g+FYQGw1Qp2JrE5AbvPJyjTeS0BDABouS+skQHy5LFnf6jJA9+9fhQnPEECWnTLv3A8gQB61Y80hLwlAVAK/VPkhxD/xbM1w6C3EPw==\",\"dtype\":\"float64\",\"shape\":[50]},\"percent_of_total\":{\"__ndarray__\":\"rFsA4pbxT0DdPz4qeJciQNlq4tjrwrs/fSC2/DhJtD+SbFzXFGOiP6fUFmfKZaA/635gSRQLij+AdPIbxgOEP20psXDL5n0/Cv/j60omdz8K/+PrSiZ3P0j/6jC4XHE/SP/qMLhccT9I/+owuFxhP8mpO1hvOFU/yak7WG84VT9I/+owuFxRP4hUk8STSkM/DVSFOrndLj9a1OqQ+FnkP3QSmIHbc5I/vZQ0tbK81D9ZUHPnFlngPyAWPp4dIMY/CRvQ+0wIuz+hEVBWPaGRP7sfkM7Uu48/ghfAHaeBdz+RFgBa8pBGPxceAHiYFi4/HGAEgv88PEAqaJDSF/YkQIXaCHC3beA/SeNFUZb3jD+syAgiKolHPzWdgzInV4g/mRbiszhQfD/r+KyuPudBP9XbRDltJE5AnN17U1+bKUCqTOt0J7MSQPgo6oaJDIk/VLyr2+OcrD9eEp4VPJeYP0hpNjhavZU/HBdnfZYJkD8SwXRDPg+EP3klUuRxXXQ/x+p0g8FKUD+ldyGfNRE6Pw==\",\"dtype\":\"float64\",\"shape\":[50]},\"percent_of_viral\":{\"__ndarray__\":\"addLqABBV0DIxHlmdRErQBgoMNqXNcQ/DJtNKvmIvT9suYRwL8WqP8I673Cx36c/T1QYQ2f1kj/2w+dY3CONP0hqJRRtxIU/1YNO5iLagD/Vg07mItqAP7/FdVk0R3k/v8V1WTRHeT+/xXVZNEdpP7Hxj/s/5V4/sfGP+z/lXj+/xXVZNEdZP7jbgio6Fkw/xq9oiC54Nj9IVz80y4cZQH3EZJ75Jcc/TizYDqUDCkBqFZ8+GoIUQDAryyGCwfs/yB1asK708D9fU7h2wh3GP9UX2WqV58M/fsT1SAN9rT8S+sJkDU98P7emLENe32I/qKLkrFmOVUDySmkBRwBAQPGeqUD6FPk/M645GMkcpj+J3a5jY/dhP5mwPt2hlKI/ttJ0VQadlT84MIAhXFVbP27seLJjFlZAUrv9o5nDMkA0LN6Ey2cbQPgZbjXvWpI/ozaJY3v3tD85rFRx+gSiP/mQF8xH3J8/BwL0nu2Alz/bFl2D72WNP4T+fjOL2H0/aZhlXG/gVz+6RlGw8hlDPw==\",\"dtype\":\"float64\",\"shape\":[50]},\"reads\":[542592,78949,921,673,305,272,108,83,62,48,48,36,36,18,11,11,9,5,2,5541,157,2823,4451,1506,920,150,135,50,6,2,4599864,1707242,83629,2304,117,1936,1126,89,1212499,257515,94027,246,1124,483,427,315,197,100,20,8],\"samples\":[\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"3_1\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"4_D\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"5_5061600092_S2\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\",\"6_A1\"],\"total_reads\":[849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,849294,871250,871250,871250,871250,871250,871250,871250,871250,871250,871250,871250,16289466,16289466,16289466,16289466,16289466,16289466,16289466,16289466,2011292,2011292,2011292,2011292,2011292,2011292,2011292,2011292,2011292,2011292,2011292,2011292],\"viral_reads\":[583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,583334,86814,86814,86814,86814,86814,86814,86814,86814,86814,86814,86814,5334770,5334770,5334770,5334770,5334770,5334770,5334770,5334770,1372384,1372384,1372384,1372384,1372384,1372384,1372384,1372384,1372384,1372384,1372384,1372384]},\"selected\":null,\"selection_policy\":null},\"id\":\"b3a90c69-a7d0-41ac-82e2-015c7c50c450\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"fill_alpha\":{\"field\":\"alphas\"},\"fill_color\":{\"field\":\"colors\"},\"height\":{\"units\":\"data\",\"value\":1},\"line_alpha\":{\"field\":\"alphas\"},\"line_color\":{\"value\":null},\"width\":{\"units\":\"data\",\"value\":1},\"x\":{\"field\":\"samples\"},\"y\":{\"field\":\"assigned\"}},\"id\":\"141687ef-33cc-42b7-a50a-12e6ccf24447\",\"type\":\"Rect\"},{\"attributes\":{\"data_source\":{\"id\":\"b3a90c69-a7d0-41ac-82e2-015c7c50c450\",\"type\":\"ColumnDataSource\"},\"glyph\":{\"id\":\"141687ef-33cc-42b7-a50a-12e6ccf24447\",\"type\":\"Rect\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"c7823b59-9ba5-4c5c-b1da-01aaca10fe11\",\"type\":\"Rect\"},\"selection_glyph\":null,\"view\":{\"id\":\"f5270b4a-ef3c-4302-9a2f-58de90b2f70c\",\"type\":\"CDSView\"}},\"id\":\"beb34f74-ecb2-45c9-9655-c7c790b4fa83\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"callback\":null,\"factors\":[\"uncultured crAssphage\",\"Streptococcus phage 315.1\",\"Sk1virus\",\"Salmonella phage SJ46\",\"Pepper mild mottle virus\",\"Parechovirus A\",\"Parabacteroides phage YZ-2015a\",\"P1virus\",\"Otarine picobirnavirus (segment 1)\",\"Norovirus\",\"Leuconostoc phage Lmd1\",\"Lactococcus virus c2\",\"Lactococcus virus bIL67\",\"Lactococcus phage bIL286\",\"Lactococcus phage Tuc2009\",\"Lactobacillus phage JCL1032\",\"Escherichia virus phiV10\",\"Escherichia virus YD2008s\",\"Escherichia virus Seurat\",\"Escherichia virus HK022\",\"Escherichia virus Cajan\",\"Escherichia virus 191\",\"Escherichia phage slur01\",\"Escherichia phage TL-2011b\",\"Escherichia phage P13374\",\"Equartevirus\",\"Enterovirus B\",\"Enterovirus A\",\"Croceibacter phage P2559Y\",\"Croceibacter phage P2559S\",\"Clostridium phage vB_CpeS-CP51\",\"Clostridium phage phiMMP04\",\"Chimpanzee faeces associated microphage 2\",\"Bacteroides phage B124-14\",\"Bacillus virus phi29\",\"Bacillus virus GA1\",\"Bacillus virus B103\",\"Bacillus phage VMY22\",\"Bacillus phage MG-B1\",\"Alphatorquevirus\",\"Actinomyces virus Av1\"]},\"id\":\"ed13e900-cb6b-461b-a049-e46e5e563b33\",\"type\":\"FactorRange\"},{\"attributes\":{},\"id\":\"62a2a7ca-677d-4c06-bb80-29f543e4d728\",\"type\":\"ResetTool\"},{\"attributes\":{\"above\":[{\"id\":\"62d6989a-f9db-41c7-9924-a4476b552845\",\"type\":\"CategoricalAxis\"}],\"left\":[{\"id\":\"1fd8bda6-c05f-4357-b404-1edbeaa36f52\",\"type\":\"CategoricalAxis\"}],\"plot_height\":720,\"plot_width\":200,\"renderers\":[{\"id\":\"62d6989a-f9db-41c7-9924-a4476b552845\",\"type\":\"CategoricalAxis\"},{\"id\":\"1ba50b14-4a35-4fbf-96df-febb9a48e254\",\"type\":\"Grid\"},{\"id\":\"1fd8bda6-c05f-4357-b404-1edbeaa36f52\",\"type\":\"CategoricalAxis\"},{\"id\":\"7ee9875a-715a-4b4c-83fd-77f0aa6f93ef\",\"type\":\"Grid\"},{\"id\":\"af50bcbe-ec5e-42eb-a370-5713d8b11bee\",\"type\":\"BoxAnnotation\"},{\"id\":\"beb34f74-ecb2-45c9-9655-c7c790b4fa83\",\"type\":\"GlyphRenderer\"}],\"title\":{\"id\":\"0d10d68d-1fdf-4e20-bc40-81fe317115d5\",\"type\":\"Title\"},\"toolbar\":{\"id\":\"86352ba3-2f64-4d3c-9e6b-83a6d1a067fa\",\"type\":\"Toolbar\"},\"x_range\":{\"id\":\"e08e0aba-3a0d-4edb-8fed-81ca2686340f\",\"type\":\"FactorRange\"},\"x_scale\":{\"id\":\"90694d91-b06e-4473-be79-72485f96d177\",\"type\":\"CategoricalScale\"},\"y_range\":{\"id\":\"ed13e900-cb6b-461b-a049-e46e5e563b33\",\"type\":\"FactorRange\"},\"y_scale\":{\"id\":\"e5fa979b-6e53-4fcd-8147-9bb839c1f534\",\"type\":\"CategoricalScale\"}},\"id\":\"ec6bf758-84e2-4a10-b53e-2f64a6ffd6fd\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"align\":\"right\",\"plot\":null,\"text\":\"GenomeDetective assignments+discoveries\",\"text_color\":{\"field\":\"-\"},\"text_font_size\":{\"value\":\"16pt\"}},\"id\":\"0d10d68d-1fdf-4e20-bc40-81fe317115d5\",\"type\":\"Title\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"height\":{\"units\":\"data\",\"value\":1},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"width\":{\"units\":\"data\",\"value\":1},\"x\":{\"field\":\"samples\"},\"y\":{\"field\":\"assigned\"}},\"id\":\"c7823b59-9ba5-4c5c-b1da-01aaca10fe11\",\"type\":\"Rect\"},{\"attributes\":{},\"id\":\"c2f857e3-6145-4b7a-b944-f39f9b9c6765\",\"type\":\"SaveTool\"},{\"attributes\":{\"callback\":null,\"factors\":[\"3_1\",\"4_D\",\"5_5061600092_S2\",\"6_A1\"]},\"id\":\"e08e0aba-3a0d-4edb-8fed-81ca2686340f\",\"type\":\"FactorRange\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"203ebb8f-3b65-4ed8-9c05-50f4c106cd0d\",\"type\":\"HoverTool\"},{\"id\":\"c2f857e3-6145-4b7a-b944-f39f9b9c6765\",\"type\":\"SaveTool\"},{\"id\":\"3a6812b9-905b-41e5-9c61-25c03fd983ed\",\"type\":\"PanTool\"},{\"id\":\"94eed387-0fd8-4070-92e9-6d9147332354\",\"type\":\"BoxZoomTool\"},{\"id\":\"a9a9c6b5-a738-4276-8223-918a56ab5148\",\"type\":\"WheelZoomTool\"},{\"id\":\"62a2a7ca-677d-4c06-bb80-29f543e4d728\",\"type\":\"ResetTool\"}]},\"id\":\"86352ba3-2f64-4d3c-9e6b-83a6d1a067fa\",\"type\":\"Toolbar\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":{\"value\":0.5},\"fill_color\":{\"value\":\"lightgrey\"},\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":{\"value\":1.0},\"line_color\":{\"value\":\"black\"},\"line_dash\":[4,4],\"line_width\":{\"value\":2},\"plot\":null,\"render_mode\":\"css\",\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"af50bcbe-ec5e-42eb-a370-5713d8b11bee\",\"type\":\"BoxAnnotation\"},{\"attributes\":{\"axis_line_color\":{\"value\":null},\"formatter\":{\"id\":\"13098a2e-4cf0-4e1d-88c9-d0f653459e15\",\"type\":\"CategoricalTickFormatter\"},\"major_label_orientation\":0.7853981633974483,\"major_label_standoff\":0,\"major_label_text_font_size\":{\"value\":\"10pt\"},\"major_tick_line_color\":{\"value\":null},\"plot\":{\"id\":\"ec6bf758-84e2-4a10-b53e-2f64a6ffd6fd\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"eba3c09b-5bd1-4c08-8649-2fa3becd2204\",\"type\":\"CategoricalTicker\"}},\"id\":\"62d6989a-f9db-41c7-9924-a4476b552845\",\"type\":\"CategoricalAxis\"},{\"attributes\":{},\"id\":\"13098a2e-4cf0-4e1d-88c9-d0f653459e15\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{\"grid_line_color\":{\"value\":null},\"plot\":{\"id\":\"ec6bf758-84e2-4a10-b53e-2f64a6ffd6fd\",\"subtype\":\"Figure\",\"type\":\"Plot\"},\"ticker\":{\"id\":\"eba3c09b-5bd1-4c08-8649-2fa3becd2204\",\"type\":\"CategoricalTicker\"}},\"id\":\"1ba50b14-4a35-4fbf-96df-febb9a48e254\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"a9a9c6b5-a738-4276-8223-918a56ab5148\",\"type\":\"WheelZoomTool\"},{\"attributes\":{},\"id\":\"1fa7cd6e-297c-42b2-b84b-9def0b534dc4\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{},\"id\":\"e5fa979b-6e53-4fcd-8147-9bb839c1f534\",\"type\":\"CategoricalScale\"},{\"attributes\":{},\"id\":\"eba3c09b-5bd1-4c08-8649-2fa3becd2204\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"source\":{\"id\":\"b3a90c69-a7d0-41ac-82e2-015c7c50c450\",\"type\":\"ColumnDataSource\"}},\"id\":\"f5270b4a-ef3c-4302-9a2f-58de90b2f70c\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"90694d91-b06e-4473-be79-72485f96d177\",\"type\":\"CategoricalScale\"}],\"root_ids\":[\"ec6bf758-84e2-4a10-b53e-2f64a6ffd6fd\"]},\"title\":\"Bokeh Application\",\"version\":\"0.12.15\"}};\n",
       "  var render_items = [{\"docid\":\"03941f5a-edb1-4dc9-995c-a7d31b56987b\",\"elementid\":\"5b4764f8-0b6f-49d8-ab26-4cbf13a6fa01\",\"modelid\":\"ec6bf758-84e2-4a10-b53e-2f64a6ffd6fd\"}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        embed_document(root);\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "      attempts++;\n",
       "      if (attempts > 100) {\n",
       "        console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\")\n",
       "        clearInterval(timer);\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "ec6bf758-84e2-4a10-b53e-2f64a6ffd6fd"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Script execution\n",
    "if __name__ == \"__main__\":\n",
    "    #Prepare dataframes\n",
    "    assignments = create_concatenated_dataframe(ASSIGNMENTS)\n",
    "    assignments[\"Assigned_Discovered\"] = \"Assigned\"\n",
    "\n",
    "    discoveries = create_concatenated_dataframe(DISCOVERIES)\n",
    "    discoveries[\"Assigned_Discovered\"] = \"Discovered\"\n",
    "\n",
    "    #Concatenate these two (assignments and discoveries)\n",
    "    results_df = pd.concat([assignments, discoveries], ignore_index=True)\n",
    "    results_df[\"run_id\"] = results_df[\"run_id\"].apply(int)\n",
    "    \n",
    "    xml_df = pd.read_csv(PARSED_XML)\n",
    "    \n",
    "    #Merge into one dataframe\n",
    "    super_df = results_df.merge(xml_df, on = [\"run_id\", \"sample_id\"], how = \"right\")\n",
    "\n",
    "    #Calculate fractions and percentages of total/viral reads\n",
    "    super_df = calculate_fractions(super_df)\n",
    "\n",
    "    #For easy use in the heatmaps, create a column that combines run_id + sample_id\n",
    "    super_df[\"sample\"] = super_df[\"run_id\"].map(str) + '_' + super_df[\"sample_id\"].map(str)\n",
    "    \n",
    "    create_heatmaps(super_df)\n",
    "    \n",
    "    #super_df.to_csv(OUTPUT_FILE, index = False)\n",
    "    #print(\"The table with all the data on which the heatmaps are based has been written to %s\" % OUTPUT_FILE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The heatmaps have been generated! Please look in the folder indicated above (in the script output) to find the interactive figures (html files)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
